Fix IsSpellCheckEnabled

jknaudt21 opened the pull request:
Note: A lot of the changed lines in this PR are just Unshipped.txt changes or tests! Don't be intimidated!!
### Problem
Editable textboxes would still spellcheck despite the property `IsSpellCheckEnabled` being set to false.

This was caused by the fact that at the Core level the class that represented editable fields, `ITextInput` , only took into consideration `IsTextPredictionEnabled` and treated it as a dual-purposed property for *both* spell checking and text suggestion (and auto correction). Conversely on the Controls level, the class that represented inputs, `InputView` only took into consideration spell checking (`IsSpellCheckEnabled`) and **not** text prediction.

All the classes that represent an editable textbox (**Entry**, **Editor**, and **SearchBar**) implemented both `InputView` and `ITextEntry`, thus they had conflicting logic  - mainly that IsSpellCheckEnabled wouldn't do anything.

### Solution 
We create parity between `InputView` and `ITextEntry` by ensuring that both classes have a way to represent spell checking (`IsSpellCheckEnabled`) and text prediction/autocomplete (`IsTextPredictionEnabled`). It is worth noting that sadly there is no perfect cross-platform matching. On Windows, enabling spellchecking on editable boxes _also_ enables auto correction (i.e Whats -> What's).  These differences, however, have been documented in the XML comments.

This in turn means adding these properties to the API, adding the `MapIsSpellCheckEnabled` mapper to the handlers, and reviewing the implementation for spellchecking & autocompletion for every platform.

For reference, here's a table that shows how each platform implements spell-checking.

| Platform |  Extended Class | Spell Checking | Text Prediction (Auto Correct) | 
| -- | -- | -- | -- | 
| Windows | `Textbox` | `Textbox.IsSpellCheckEnabled` both spellchecking and autocorrection with this setting due to platform limitation | `Textbox.IsTextPredictionEnabled` shows keyboard suggestions  |
| iOS | `TextField` | `Textfield.SpellCheckingType` set to a value of the `UITextSpellCheckingType` enum | `Textfield.AutocorrectionType` set to a value of `UITextAutocorrectionType` enum | 
| Android | `EditText` | `EditText.InputType` does a bitwise inclusion of the `InputTypes.TextFlagNoSuggestions` enum value | `EditText.InputType` does a bitwise inclusion of the `InputTypes.TextFlagAutoCorrect` enum value |

### Demo
Result of typing "Whats new, frend". When TextPrediction is enabled, we should see "Whats" being turned to "What's"
**Android**
![image](https://github.com/dotnet/maui/assets/32918747/8555a519-40d7-44c3-b146-71540c496cd7)
**Windows** ([text prediction only kicks in using touch keyboards](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox?view=winrt-22621#use-a-text-box-with-the-touch-keyboard) ). Also as noted before, Window's won't correct "Whats" to "What's" unless spell check is turned on
![image](https://github.com/dotnet/maui/assets/32918747/97dafacf-ffdc-4f89-a909-b410ea9d3019).

**iOS**
Note: iOS/MacCatalyst doesn't count "Whats" as word that's spelled wrong, but will autocorrect you if Text prediction is on
<img width="450" alt="image" src="https://github.com/dotnet/maui/assets/32918747/65cbdab7-5fe3-4572-b0f5-6074747d2cd3">


### Issues Fixed

Fixes #8558 

### Future Work
I noticed that the code that pertains to special keyboards is arbitrary and there is little to no clarity on what should the keyboards be doing. I plan on submitting (and fixing) an issue in the near future once this PR is reviewed that fully standardizes keyboard behavior as well as documenting the properties so people understand the differences.

jknaudt21 committed:
Change API docs

jknaudt21 committed:
Tentative changes

jknaudt21 committed:
Add MapIsSpellCheckEnabled mapper

This at the xplat layer. Implementations will come later

jknaudt21 committed:
Add iOS IsSpellCheckEnabled mapper

jknaudt21 committed:
Add Windows IsSpellCheckEnabled mapper

jknaudt21 committed:
Add Android IsSpellCheckEnabled mapper

jknaudt21 committed:
Add Windows searchbar mapper

jknaudt21 committed:
Add iOS searchbar mapper

jknaudt21 committed:
Add Android searchbar mapper

jknaudt21 committed:
Add Windows editor mapper

jknaudt21 committed:
Add iOS editor mapper

jknaudt21 committed:
Add Android editor mapper

jknaudt21 committed:
Add API changes to Unshipped.txt

jknaudt21 committed:
Remove unecessary usings

jknaudt21 committed:
More RS0016

jknaudt21 committed:
More RS0016

jknaudt21 committed:
Revise Android inputs &  other small errors

jknaudt21 committed:
Add IsSpellCheckEnabled to test stubs

jknaudt21 committed:
Add Tizen mappers

jknaudt21 committed:
Add xml comment

jknaudt21 committed:
More tizen build fixes

jknaudt21 committed:
More tizen RS0016

jknaudt21 committed:
More tizen RS0016

jknaudt21 committed:
Start work tests (missing Keyboard-related tests)

jknaudt21 committed:
Fix iOS typo

jknaudt21 committed:
Windows and Android test checkin

jknaudt21 committed:
Fix iOS typo

jknaudt21 committed:
Checkin iOS tests

jknaudt21 committed:
Finish Tests!

jknaudt21 committed:
Clear up documentation

jknaudt21 commented on src/Controls/docs/Microsoft.Maui.Controls/InputView.xml:
```diff
@@ -97,46 +97,6 @@
         <remarks>To be added.</remarks>
       </Docs>
     </Member>
-    <Member MemberName="IsSpellCheckEnabled">
```
These comments were moved to be inline XML (since that's the way we're documenting now)

jknaudt21 commented on src/Core/src/Platform/Android/KeyboardExtensions.cs:
```diff
@@ -37,37 +30,16 @@ public static InputTypes ToInputType(this Keyboard self)
 				var spellcheckEnabled = (custom.Flags & KeyboardFlags.Spellcheck) == KeyboardFlags.Spellcheck;
 				var suggestionsEnabled = (custom.Flags & KeyboardFlags.Suggestions) == KeyboardFlags.Suggestions;
 
-				if (!capitalizedSentenceEnabled && !spellcheckEnabled && !suggestionsEnabled)
```
This block of "if" statements went through every possible combination of `capitalizedSentenceEnabled`, `spellcheckEnabled`, `suggestionsEnabled`. I simplified it by just turning on the appropiat flags.

jknaudt21 commented on src/Controls/src/Core/Platform/Windows/Extensions/AutoSuggestBoxExtensions.cs:
```diff
@@ -6,16 +6,6 @@ namespace Microsoft.Maui.Controls.Platform
 {
 	internal static class AutoSuggestBoxExtensions
 	{
-		public static void UpdateIsSpellCheckEnabled(this AutoSuggestBox platformControl, SearchBar searchBar)
```
Moved to Windows SearchBar extensions

jknaudt21 commented on src/Core/src/Platform/Windows/TextBoxExtensions.cs:
```diff
@@ -169,7 +174,7 @@ internal static void UpdateInputScope(this TextBox textBox, ITextInput textInput
 			else
 			{
 				textBox.IsTextPredictionEnabled = textInput.IsTextPredictionEnabled;
-				textBox.IsSpellCheckEnabled = textInput.IsTextPredictionEnabled;
```
This was the main source of the bug that customers were reporting! :)

jknaudt21 committed:
Merge branch 'main' into dev/jd/entry

jknaudt21 committed:
Fix isLoaded typo

jknaudt21 committed:
Merge branch 'main' into dev/jd/entry

PureWeen commented on src/Controls/src/Core/Entry/Entry.Controls.cs:
```diff
@@ -79,11 +79,6 @@ public partial class Entry : InputView, IFontElement, ITextAlignmentElement, IEn
-		/// <summary>
-		/// Backing store for the <see cref="IsTextPredictionEnabled"/> property.
-		/// </summary>
-		public static readonly BindableProperty IsTextPredictionEnabledProperty = BindableProperty.Create(nameof(IsTextPredictionEnabled), typeof(bool), typeof(Entry), true, BindingMode.OneTime);
```
I think moving this will break binary compatibility. 

@mattleibow thoughts?

PureWeen replied:
Here's an example of what we've done for this in the past

https://github.com/xamarin/Xamarin.Forms/pull/4477/files#diff-50d6e1619991945a6507073b90fef55b04c6ba125d3da2b13fd5800b6997d4aeL20-L22

mattleibow replied:
Yep. However, we also need to keep the plain properties as well - otherwise people upgrading from net7 to net8 will suddenly get property not found exceptions. Just use the `new` and call into the same property. A bit annoying and looks weird, but that is just us - the rest of the devs are happy their code still works.

jknaudt21 replied:
@mattleibow  so do you mean something like this?
```c#
public static new readonly BindableProperty IsTextPredictionEnabledProperty = InputView.IsTextPredictionEnabledProperty;
``` 

Also, would I need to bring back this as well? 
```c#
public bool IsTextPredictionEnabled
{
    get { return (bool)GetValue(IsTextPredictionEnabledProperty); }
    set { SetValue(IsTextPredictionEnabledProperty, value); }
}
```

jknaudt21 replied:
Added this commit: https://github.com/dotnet/maui/compare/38eae6033d4b0e5d91f5d6fb324a74549a77bef4...7ab5dac808fd9efe7b84a6576a42c0d41e6455ed. Let me know if this was all that was missing :)

jknaudt21 committed:
Undo binary breaks

jknaudt21 committed:
Merge branch 'main' into dev/jd/entry

jknaudt21 committed:
Undo more binary breaks

jknaudt21 committed:
Merge branch 'main' into dev/jd/entry

jfversluis reviewed (commented):
Some minor (comment) things. I think the code in itself looks good!

jknaudt21 committed:
Address PR comments

Co-authored-by: Gerald Versluis <gerald.versluis@microsoft.com>

jknaudt21 committed:
Add Mapper documentation

Adding only docs for the standard implementation. A brief search into the handler folder showed that no mapper was documented 🥲

I'm hoping that documenting the standard implementation will be what's displayed in the MSLearn page

jknaudt21 commented on src/Core/src/Handlers/Editor/EditorHandler.Android.cs:
```diff
@@ -90,6 +90,9 @@ public static void MapPlaceholderColor(IEditorHandler handler, IEditor editor)
 		public static void MapIsTextPredictionEnabled(IEditorHandler handler, IEditor editor) =>
 			handler.PlatformView?.UpdateIsTextPredictionEnabled(editor);
 
+		public static void MapIsSpellCheckEnabled(IEditorHandler handler, IEditor editor) =>
```
@jfversluis Docs added on this commit! https://github.com/dotnet/maui/pull/15459/commits/c6defcc0b83a455bd9aac2d7f2170851fbee9bad 
I'm hoping that documenting the Standard implementation is sufficient to have it reflect on MSLearn and other places, though its hard to tell since I couldn't find a documented mapper in the Handlers folder 🥲

jsuarezruiz committed:
Updated samples

PureWeen commented on src/Controls/src/Core/Entry/Entry.cs:
```diff
@@ -179,15 +179,6 @@ public bool FontAutoScalingEnabled
-		/// <summary>
-		/// Determines whether text prediction and automatic text correction is enabled. Default value is <see langword="true"/>.
-		/// </summary>
-		public bool IsTextPredictionEnabled
```
@mattleibow @mandel-macaque is this a binary break? 
The PR moves this property up to the base class.

mandel-macaque replied:
As long as the base class is in the same binary, this is not a breaking change, otherwise, it is a binary break.

jknaudt21 replied:
src/Compatibility/Material/src/Tizen/Compatibility.Material.Tizen.csproj  references these classes so yeah ... its a break. I'll get on to fixing it

jknaudt21 replied:
Seems we're on the clear! Everything is inside controls :D

jknaudt21 commented:
There seems to be a problem with Android `IsTextPredictionEnabled`, where it is only taking effect if `IsSpellCheckEnabled = false` 

```
        <VerticalStackLayout>
            <Label Text="1. Enter the string 'hi my name is james' exactly (case-sensitive)." />
            <Label Text="2. The test fails if the keyboard presents word suggestions." />
            <Editor IsTextPredictionEnabled="False" IsSpellCheckEnabled="False" />

            <Label Text="3. Enter the string 'hi my name is james' exactly (case-sensitive).( />
            <Label Text="4. The test fails if the keyboard does not present word suggestions." />
            <Editor IsTextPredictionEnabled="False" IsSpellCheckEnabled="True"/>
        </VerticalStackLayout>
```

![repro2](https://github.com/dotnet/maui/assets/32918747/32a0a95f-474e-45bf-a556-105d9d20cc5c)

On the gif above, one would expect the second textbox to underline the mispelled words "namev" and "james" **but not** auto-correct them :( 

Linked to #17690
